---
title: "The Kiel Microbiome Center workshop"
---

# Overview

Here you can find the teaching and practical materials from the workshop.

## Concepts in microbiome ecology and data analysis

Link all files of the talks here, use a logical order:

1.  [Data processing](talks/nfdi4microbiota_DataPreProcessing.pdf)
2.  [Alpha diversity](talks/nfdi4microbiota_Alpha_diversity.pdf)
3.  [Beta diversity](talks/nfdi4microbiota_Beta_diversity.pdf)
4.  [Metagenomics](talks/nfdi4microbiota_DAA_Metagenomics.pdf)

## Hands-on material

1.  [Introduction to R](scripts/1_intro_to_r.ipynb)
2.  [Alpha diversity analysis](scripts/2_Alpha_diversity.ipynb)
3.  [Beta diversity analysis](scripts/3_Beta_diversity.ipynb)
4.  [Differential abundance analysis](scripts/4_Differential_abundance.ipynb)
5.  [Metagenome outlook](scripts/5_Metagenome_outook.ipynb)

Commands are in R. Scripts are in Jupyter notebook. Below, we provide instructions on how to set it all up.

# Setting up the tutorial

Download the whole [project directory](https://github.com/ikmb/nfdi4microbiota_kmc_microbiome_workshop/archive/refs/heads/main.zip) to your computer. Unzip the folder. Scripts are in [`scripts` folder](https://github.com/ikmb/nfdi4microbiota_kmc_microbiome_workshop/tree/main/scripts) and inputs are in the [`inputs` folder](https://github.com/ikmb/nfdi4microbiota_kmc_microbiome_workshop/tree/main/inputs).

To run the tutorial, it is assumed that you are working on a mac or a linux-based computer.

## Input files

### 16S Data Files

To run the tutorials, you will need files from the [`inputs` folder](https://github.com/ikmb/nfdi4microbiota_kmc_microbiome_workshop/tree/main/inputs) with 16S dataset. This dataset has been reduced in size and randomly shuffled compared to the original dataset. The files you need are:

-   feature-table.tsv
-   metadata_bn10_complete_fixed.tsv
-   taxonomy_form.csv
-   tree.nwk

### Metagenome Files:

For metagenome analysis, you will need to download the files from public repositories and place them in the `inputs` folder.

#### HMP Sample

These files are to be used in the tutorial a hmp2 metagenome with the ID MSM79H8.

-   [SRR5935872_1.fastq.gz](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR593/002/SRR5935872/SRR5935872_1.fastq.gz)
-   [SRR5935872_2.fastq.gz](http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR593/002/SRR5935872/SRR5935872_2.fastq.gz)

The files need to be downloaded at (`inputs/metagenome/`).

#### Human/GRC38 genome files

-   genome.1.bt2
-   genome.2.bt2
-   genome.3.bt2
-   genome.4.bt2
-   genome.fa
-   genome.rev.1.bt2
-   genome.rev.2.bt2

You can download the needed Human/GRC38 genome files from e.g. [here](https://genome-idx.s3.amazonaws.com/bt/GRCh38_noalt_decoy_as.zip). Unzip the files at `inputs/metagenome/`.


#### Sylph database

Download the file to (`inputs/metagenome/`).

-   [gtdb-r220-c200-dbv1.syldb](http://faust.compbio.cs.cmu.edu/sylph-stuff/gtdb-r220-c200-dbv1.syldb)

The used database can be downloaded from the [official Sylph Repository](https://github.com/bluenote-1577/sylph/wiki/Pre%E2%80%90built-databases). Note that this file has around 13 Gibabytes.


#### GTDB-Tk taxonomy

Download the file to (`inputs/metagenome/`).

-  [bac120_taxonomy_r220.tsv.gz](https://data.ace.uq.edu.au/public/gtdb/data/releases/release220/220.0/bac120_taxonomy_r220.tsv.gz)

#### BBMap Files

Download the files to (`inputs/metagenome/bbmap/`).

-   nextera.fa.gz
-   phix174_ill.ref.fa.gz
-   phix_adapters.fa.gz

You can download the needed files from the [BBMap Repository](https://github.com/BioInfoTools/BBMap/tree/master/resources).

#### Intermediary back-up files for phyloseq

These files are intermediary files. You will create these in the tutorial. Many of the tutorial use these objects as input. Therefore, we are providing them here.

!!!!!! TheseCreate additionally a directory called `R_objects`, which will contain:

-   full_phyloseq_object.RData
-   rare8000_phyloseq_object.RData


# Running the R commands

## Option 1: Using Jupyter Notebooks with the R Kernel

### Install conda environment

First, make sure you have conda installed. You can install all required packages and tools with [conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/linux.html). Once you install conda, you can create a dedicated conda environment for the tutorial with the following command:

``` bash
conda env create --file kmc_16s_environment.yml --prefix kmc_workshop
```

Once the environment is created, activate it:

``` bash
conda activate kmc_workshop
```

### Launch Jupyter Notebook

If Jupyter is not already installed, you can install it within the environment:

``` bash
conda install jupyter
```

Start Jupyter with the R kernel:

``` bash
jupyter notebook
```

## Option 2: Using RStudio

Ensure that conda is installed and active (same as step 1 above).

#### Link Conda Environment with RStudio:

In RStudio, go to Tools \> Global Options \> R and select the R Version associated with the Conda environment.

#### Run Your R Code

You can now open your R scripts or RMarkdown notebooks in RStudio and run the code, just as we did in the Jupyter environment.
